### PR TITLE
feat: refactor http calls to ModuleApi

### LINKED ISSUE #31: Outgoing requests should use the Synapse HTTP client exposed via the module API
The synapse module API exposes a [SimpleHTTPClient](https://github.com/famedly/synapse/blob/master/synapse/http/client.py#L754) on `ModuleApi.http_client`. We currently just use `requests` for making HTTP requests, we should refactor the existing code to use the synapse-provided HTTP client.

### PR DESCRIPTION
**Pull request by sirewix**
Closes #31

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -23,14 +23,21 @@ classifiers = [
   "Programming Language :: Python :: Implementation :: CPython",
   "Programming Language :: Python :: Implementation :: PyPy",
 ]
-dependencies = ["jwcrypto", "twisted"]
+dependencies = [
+  "jwcrypto",
+  "twisted",
+  "synapse @ git+https://github.com/famedly/synapse.git@v1.103.0_1",
```
Comment (nico-famedly) [inline pyproject.toml:29]: This is the wrong package name, but also you don't really want to depend on synapse directly. That has caused problems in the past, see https://github.com/famedly/synapse-invite-checker/commit/858a6ed6e856d5cde971e1013ad2cfad0d68a0a8

File: pyproject.toml [hunk 2]
```diff
@@ -31,6 +31,9 @@ Documentation = "https://github.com/famedly/synapse-token-authenticator"
 Issues = "https://github.com/famedly/synapse-token-authenticator/issues"
 Source = "https://github.com/famedly/synapse-token-authenticator"
 
+[tool.hatch.metadata]
+allow-direct-references = true
+
```
Comment (nico-famedly) [inline pyproject.toml:36]: This seems to be not needed anymore, I think?

**Review by tlater-famedly — DISMISSED** [review]
Generally fine, but more granular commits would be appreciated in future.

I can't really comment on the tests, @sirewix has asked @nico-famedly for help there, so I'll leave that part to them :)

File: pyproject.toml [hunk 3]
```diff
@@ -6,7 +6,7 @@ build-backend = "hatchling.build"
 name = "synapse-token-authenticator"
 description = 'Synapse authentication module which allows for authenticating and registering using JWTs'
 readme = "README.md"
-requires-python = ">=3.9"
+requires-python = ">=3.11"
```
[the remark this answers is not in the document]
Comment (tlater-famedly) [reply pyproject.toml:9]: Looks like our containers run 3.11, so this is probably fine: https://github.com/famedly/synapse/blob/53d399896c527688e1a4b6c349d0306ddf4ad228/docker/Dockerfile#L23

File: tests/__init__.py [hunk 4]
```diff
@@ -14,65 +13,127 @@
...
+        cls._patcher4.stop()
+
+    servlets = [
+        admin.register_servlets,
+        account_data.register_servlets,
+        login.register_servlets,
+        room.register_servlets,
+        presence.register_servlets,
+        profile.register_servlets,
+        notifications.register_servlets,
```
Comment (nico-famedly) [inline tests/__init__.py:93]: Nitpick, but you probably don't need all of these?

File: synapse_token_authenticator/utils.py [hunk 5]
```diff
@@ -1,5 +1,7 @@
-import requests
+from base64 import b64encode
 from urllib.parse import urljoin
+
+import requests
```
Comment (nico-famedly) [inline synapse_token_authenticator/utils.py:4]: There is still one requests call. If that can't be changed to the builtin client, could you document why? (A comment or so)

**Review by nico-famedly — COMMENTED** [review]
Looks mostly good to me and like I would expect it to work, but I could be missing stuff :)

File: synapse_token_authenticator/utils.py (hunk 5 above)
Comment (sirewix) [reply synapse_token_authenticator/utils.py:4]: Totally missed this one, thx

Comment (sirewix) [reply synapse_token_authenticator/utils.py:4]: Huh, this function seems to be unused

https://github.com/famedly/synapse-token-authenticator/blob/ba1c267f4d7e449ce3a28a68ec1ded25a9ce3e53/synapse_token_authenticator/utils.py#L27-L34

Should we remove it?

**Comment by emgrav** [comment]
I agree with Tristan that practicing good commit hygiene is important and something to keep in mind for future work, but I think this is good for now. There are some deprecation warnings in the 3.11 test output that would probably be good to take care of right away, but otherwise LGTM.

**Review by tlater-famedly — APPROVED** [review]
@emgrav reviewed the latest state, GitHub discussion UI and our attempt at fixing it seems to have caused that to be lost.

**Review by tlater-famedly — APPROVED** [review]
LGTM - we should consider adding these linters to the repo as a GitHub action, otherwise things will probably drift in the future. But that should be a separate topic.

**Pull request merged**